In [1]:
import numpy as np
import pandas as pd

In [7]:
# numpy version check
print(np.__version__)
# Pandas version
print(pd.__version__)

1.16.5
0.25.1


In [3]:
# Read the data
data = pd.read_csv("FL_insurance_sample.csv")

In [4]:
# Checking output type
print(type(data))

<class 'pandas.core.frame.DataFrame'>


In [6]:
# To check the shape 
print(data.shape)
# to get the number of rows or instances or records or examples
print(data.shape[0])
# to get the number of columns or features or attributes
print(data.shape[1])

(36634, 18)
36634
18


In [9]:
# To get the column names
print(data.columns)
# To get the row names
print(data.index)

Index(['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit',
       'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012',
       'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible',
       'fr_site_deductible', 'point_latitude', 'point_longitude', 'line',
       'construction', 'point_granularity'],
      dtype='object')
RangeIndex(start=0, stop=36634, step=1)


In [14]:
# Datatype of each column
print(data.dtypes)
print("-----"*15)
# what is output type for data.dtypes
print(type(data.dtypes))
print("-----"*15)
# Access the first column dtype
print(data.dtypes[0])
print("-----"*15)
# Access the last column dtype
print(data.dtypes[-1])

policyID                int64
statecode              object
county                 object
eq_site_limit         float64
hu_site_limit         float64
fl_site_limit         float64
fr_site_limit         float64
tiv_2011              float64
tiv_2012              float64
eq_site_deductible    float64
hu_site_deductible    float64
fl_site_deductible    float64
fr_site_deductible      int64
point_latitude        float64
point_longitude       float64
line                   object
construction           object
point_granularity       int64
dtype: object
---------------------------------------------------------------------------
<class 'pandas.core.series.Series'>
---------------------------------------------------------------------------
int64
---------------------------------------------------------------------------
int64


In [28]:
# Column count for each datatype
# How many columns are there with int64
# How many columns are there with float64
# How many columns are there with object
data.dtypes.value_counts()
# Note We are using series method value_counts

float64    11
object      4
int64       3
dtype: int64

##### observe the data

In [15]:
# display top 5 rows
data.head()

policyID statecode       county  eq_site_limit  hu_site_limit  \
0    119736        FL  CLAY COUNTY       498960.0      498960.00   
1    448094        FL  CLAY COUNTY      1322376.3     1322376.30   
2    206893        FL  CLAY COUNTY       190724.4      190724.40   
3    333743        FL  CLAY COUNTY            0.0       79520.76   
4    172534        FL  CLAY COUNTY            0.0      254281.50   

   fl_site_limit  fr_site_limit    tiv_2011    tiv_2012  eq_site_deductible  \
0       498960.0       498960.0   498960.00   792148.90                 0.0   
1      1322376.3      1322376.3  1322376.30  1438163.57                 0.0   
2       190724.4       190724.4   190724.40   192476.78                 0.0   
3            0.0            0.0    79520.76    86854.48                 0.0   
4            0.0       254281.5   254281.50   246144.49                 0.0   

   hu_site_deductible  fl_site_deductible  fr_site_deductible  point_latitude  \
0              9979.2                 0.0                   0       30.102261   
1                 0.0                 0.0                   0       30.063936   
2                 0.0                 0.0                   0       30.089579   
3                 0.0                 0.0                   0       30.063236   
4                 0.0                 0.0                   0       30.060614   

   point_longitude         line construction  point_granularity  
0       -81.711777  Residential      Masonry                  1  
1       -81.707664  Residential      Masonry                  3  
2       -81.700455  Residential         Wood                  1  
3       -81.707703  Residential         Wood                  3  
4       -81.702675  Residential         Wood                  1

In [16]:
# display top 10 rows
data.head(10)

policyID statecode       county  eq_site_limit  hu_site_limit  \
0    119736        FL  CLAY COUNTY       498960.0      498960.00   
1    448094        FL  CLAY COUNTY      1322376.3     1322376.30   
2    206893        FL  CLAY COUNTY       190724.4      190724.40   
3    333743        FL  CLAY COUNTY            0.0       79520.76   
4    172534        FL  CLAY COUNTY            0.0      254281.50   
5    785275        FL  CLAY COUNTY            0.0      515035.62   
6    995932        FL  CLAY COUNTY            0.0    19260000.00   
7    223488        FL  CLAY COUNTY       328500.0      328500.00   
8    433512        FL  CLAY COUNTY       315000.0      315000.00   
9    142071        FL  CLAY COUNTY       705600.0      705600.00   

   fl_site_limit  fr_site_limit     tiv_2011     tiv_2012  eq_site_deductible  \
0       498960.0       498960.0    498960.00    792148.90                 0.0   
1      1322376.3      1322376.3   1322376.30   1438163.57                 0.0   
2       190724.4       190724.4    190724.40    192476.78                 0.0   
3            0.0            0.0     79520.76     86854.48                 0.0   
4            0.0       254281.5    254281.50    246144.49                 0.0   
5            0.0            0.0    515035.62    884419.17                 0.0   
6            0.0            0.0  19260000.00  20610000.00                 0.0   
7       328500.0       328500.0    328500.00    348374.25                 0.0   
8       315000.0       315000.0    315000.00    265821.57                 0.0   
9       705600.0       705600.0    705600.00   1010842.56             14112.0   

   hu_site_deductible  fl_site_deductible  fr_site_deductible  point_latitude  \
0              9979.2                 0.0                   0       30.102261   
1                 0.0                 0.0                   0       30.063936   
2                 0.0                 0.0                   0       30.089579   
3                 0.0                 0.0                   0       30.063236   
4                 0.0                 0.0                   0       30.060614   
5                 0.0                 0.0                   0       30.063236   
6                 0.0                 0.0                   0       30.102226   
7             16425.0                 0.0                   0       30.102217   
8             15750.0                 0.0                   0       30.118774   
9             35280.0                 0.0                   0       30.100628   

   point_longitude         line         construction  point_granularity  
0       -81.711777  Residential              Masonry                  1  
1       -81.707664  Residential              Masonry                  3  
2       -81.700455  Residential                 Wood                  1  
3       -81.707703  Residential                 Wood                  3  
4       -81.702675  Residential                 Wood                  1  
5       -81.707703  Residential              Masonry                  3  
6       -81.713882   Commercial  Reinforced Concrete                  1  
7       -81.707146  Residential                 Wood                  1  
8       -81.704613  Residential                 Wood                  1  
9       -81.703751  Residential              Masonry                  1

In [18]:
# display last 5 rows
data.tail()

policyID statecode           county  eq_site_limit  hu_site_limit  \
36629    404309        FL  PINELLAS COUNTY            0.0        88722.0   
36630    465875        FL  PINELLAS COUNTY      1297057.5      1297057.5   
36631    791209        FL  PINELLAS COUNTY       173286.9       173286.9   
36632    322627        FL  PINELLAS COUNTY      1499781.6      1499781.6   
36633    398149        FL  PINELLAS COUNTY       373488.3       373488.3   

       fl_site_limit  fr_site_limit   tiv_2011    tiv_2012  \
36629            0.0        88722.0    88722.0   155065.65   
36630      1297057.5      1297057.5  1297057.5  1577500.79   
36631            0.0            0.0   173286.9   205527.97   
36632            0.0            0.0  1499781.6  1768782.43   
36633            0.0            0.0   373488.3   596003.67   

       eq_site_deductible  hu_site_deductible  fl_site_deductible  \
36629                 0.0                 0.0                 0.0   
36630                 0.0                 0.0                 0.0   
36631                 0.0                 0.0                 0.0   
36632                 0.0                 0.0                 0.0   
36633                 0.0                 0.0                 0.0   

       fr_site_deductible  point_latitude  point_longitude         line  \
36629                   0       28.122885       -82.770218  Residential   
36630                   0       28.080900       -82.758800  Residential   
36631                   0       28.110550       -82.766360  Residential   
36632                   0       28.089415       -82.697411  Residential   
36633                   0       28.064440       -82.774590  Residential   

      construction  point_granularity  
36629         Wood                  3  
36630      Masonry                  1  
36631         Wood                  4  
36632      Masonry                  3  
36633      Masonry                  1

In [19]:
# display last 3 rows
data.tail(3)

policyID statecode           county  eq_site_limit  hu_site_limit  \
36631    791209        FL  PINELLAS COUNTY       173286.9       173286.9   
36632    322627        FL  PINELLAS COUNTY      1499781.6      1499781.6   
36633    398149        FL  PINELLAS COUNTY       373488.3       373488.3   

       fl_site_limit  fr_site_limit   tiv_2011    tiv_2012  \
36631            0.0            0.0   173286.9   205527.97   
36632            0.0            0.0  1499781.6  1768782.43   
36633            0.0            0.0   373488.3   596003.67   

       eq_site_deductible  hu_site_deductible  fl_site_deductible  \
36631                 0.0                 0.0                 0.0   
36632                 0.0                 0.0                 0.0   
36633                 0.0                 0.0                 0.0   

       fr_site_deductible  point_latitude  point_longitude         line  \
36631                   0       28.110550       -82.766360  Residential   
36632                   0       28.089415       -82.697411  Residential   
36633                   0       28.064440       -82.774590  Residential   

      construction  point_granularity  
36631         Wood                  4  
36632      Masonry                  3  
36633      Masonry                  1

In [20]:
# select a random row
data.sample()

policyID statecode      county  eq_site_limit  hu_site_limit  \
31823    171221        FL  LEE COUNTY            0.0      123673.48   

       fl_site_limit  fr_site_limit   tiv_2011   tiv_2012  eq_site_deductible  \
31823            0.0            0.0  123673.48  108278.61                 0.0   

       hu_site_deductible  fl_site_deductible  fr_site_deductible  \
31823                 0.0                 0.0                   0   

       point_latitude  point_longitude         line construction  \
31823       26.632211       -81.849019  Residential         Wood   

       point_granularity  
31823                  1

In [21]:
# select 5 rows randomly
data.sample(5)

policyID statecode             county  eq_site_limit  hu_site_limit  \
15756    973925        FL  MIAMI DADE COUNTY            0.0      170249.04   
30092    136646        FL   HIGHLANDS COUNTY            0.0     1097216.10   
20330    193183        FL  MIAMI DADE COUNTY      4166100.0     4189500.00   
6281     181430        FL  WASHINGTON COUNTY            0.0       25002.92   
11254    543315        FL     VOLUSIA COUNTY            0.0     4839381.00   

       fl_site_limit  fr_site_limit    tiv_2011    tiv_2012  \
15756            0.0            0.0   170249.04   221304.68   
30092            0.0      1097216.1  1097216.10  1251892.85   
20330      4166100.0      4166100.0  4189500.00  3946408.45   
6281             0.0            0.0    25002.92    26643.11   
11254            0.0            0.0  4839381.00  5595640.75   

       eq_site_deductible  hu_site_deductible  fl_site_deductible  \
15756                 0.0                 0.0                 0.0   
30092                 0.0                 0.0                 0.0   
20330                 0.0            209475.0                 0.0   
6281                  0.0                 0.0                 0.0   
11254                 0.0                 0.0                 0.0   

       fr_site_deductible  point_latitude  point_longitude         line  \
15756                   0       25.876320       -80.306206  Residential   
30092                   0       27.566740       -81.503400  Residential   
20330                   0       25.805304       -80.341675   Commercial   
6281                    0       30.795090       -85.536797  Residential   
11254                   0       29.031698       -81.308990   Commercial   

             construction  point_granularity  
15756                Wood                  1  
30092             Masonry                  1  
20330  Reinforced Masonry                  1  
6281                 Wood                  1  
11254  Reinforced Masonry                  1

- data.head()
- data.tail()
- data.samle()

the output type of all the above methods are dataframe

In [23]:
data.head(1)

policyID statecode       county  eq_site_limit  hu_site_limit  \
0    119736        FL  CLAY COUNTY       498960.0       498960.0   

   fl_site_limit  fr_site_limit  tiv_2011  tiv_2012  eq_site_deductible  \
0       498960.0       498960.0  498960.0  792148.9                 0.0   

   hu_site_deductible  fl_site_deductible  fr_site_deductible  point_latitude  \
0              9979.2                 0.0                   0       30.102261   

   point_longitude         line construction  point_granularity  
0       -81.711777  Residential      Masonry                  1

In [24]:
data.tail(1)

policyID statecode           county  eq_site_limit  hu_site_limit  \
36633    398149        FL  PINELLAS COUNTY       373488.3       373488.3   

       fl_site_limit  fr_site_limit  tiv_2011   tiv_2012  eq_site_deductible  \
36633            0.0            0.0  373488.3  596003.67                 0.0   

       hu_site_deductible  fl_site_deductible  fr_site_deductible  \
36633                 0.0                 0.0                   0   

       point_latitude  point_longitude         line construction  \
36633        28.06444        -82.77459  Residential      Masonry   

       point_granularity  
36633                  1

- data.head(1)
- data.samle(1)

the output type of all the above methods are also  dataframe which consists a single row

#### Info about the dataframe

In [22]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36634 entries, 0 to 36633
Data columns (total 18 columns):
policyID              36634 non-null int64
statecode             36634 non-null object
county                36634 non-null object
eq_site_limit         36634 non-null float64
hu_site_limit         36634 non-null float64
fl_site_limit         36634 non-null float64
fr_site_limit         36634 non-null float64
tiv_2011              36634 non-null float64
tiv_2012              36634 non-null float64
eq_site_deductible    36634 non-null float64
hu_site_deductible    36634 non-null float64
fl_site_deductible    36634 non-null float64
fr_site_deductible    36634 non-null int64
point_latitude        36634 non-null float64
point_longitude       36634 non-null float64
line                  36634 non-null object
construction          36634 non-null object
point_granularity     36634 non-null int64
dtypes: float64(11), int64(3), object(4)
memory usage: 5.0+ MB


In [26]:
type(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36634 entries, 0 to 36633
Data columns (total 18 columns):
policyID              36634 non-null int64
statecode             36634 non-null object
county                36634 non-null object
eq_site_limit         36634 non-null float64
hu_site_limit         36634 non-null float64
fl_site_limit         36634 non-null float64
fr_site_limit         36634 non-null float64
tiv_2011              36634 non-null float64
tiv_2012              36634 non-null float64
eq_site_deductible    36634 non-null float64
hu_site_deductible    36634 non-null float64
fl_site_deductible    36634 non-null float64
fr_site_deductible    36634 non-null int64
point_latitude        36634 non-null float64
point_longitude       36634 non-null float64
line                  36634 non-null object
construction          36634 non-null object
point_granularity     36634 non-null int64
dtypes: float64(11), int64(3), object(4)
memory usage: 5.0+ MB


NoneType

#### Accesing a single column

In [30]:
# To access the column policyID
data.policyID

0        119736
1        448094
2        206893
3        333743
4        172534
          ...  
36629    404309
36630    465875
36631    791209
36632    322627
36633    398149
Name: policyID, Length: 36634, dtype: int64

In [32]:
# What is the ouput type
print(type(data.policyID))

<class 'pandas.core.series.Series'>


In [31]:
# To access the column policyID
data['policyID']

0        119736
1        448094
2        206893
3        333743
4        172534
          ...  
36629    404309
36630    465875
36631    791209
36632    322627
36633    398149
Name: policyID, Length: 36634, dtype: int64

In [33]:
# What is the ouput type
print(type(data['policyID']))

<class 'pandas.core.series.Series'>


##### specific column Datatype

In [34]:
data.statecode

0        FL
1        FL
2        FL
3        FL
4        FL
         ..
36629    FL
36630    FL
36631    FL
36632    FL
36633    FL
Name: statecode, Length: 36634, dtype: object

In [35]:
# Specific column datatype
data.statecode.dtype

dtype('O')

In [36]:
data.statecode.dtypes

dtype('O')

In [37]:
data.dtype

AttributeError: 'DataFrame' object has no attribute 'dtype'

difference b/n dtype and dtypes
- datatafrme object doesn't have dtype

- Numerical columns - int or float data
    - mean,median,mode,sum.quartiles,min,max
- Categorical columns - Object type(string data)
- Bollean data - True or False

In [40]:
data.point_granularity

0        1
1        3
2        1
3        3
4        1
        ..
36629    3
36630    1
36631    4
36632    3
36633    1
Name: point_granularity, Length: 36634, dtype: int64

In [41]:
data.point_granularity.dtype

dtype('int64')

In [42]:
# mean value of a column
data.point_granularity.mean()

1.6409073538243162

In [43]:
# mean value of a column
data.point_granularity.sum()/data.point_granularity.shape[0]

1.6409073538243162

In [45]:
# minimum value of a column
print(data.point_granularity.min())
# maximum value of a column
print(data.point_granularity.max())

1
7


In [47]:
data.point_granularity.value_counts()

1    26569
3     6724
4     3337
7        2
5        2
Name: point_granularity, dtype: int64

In [46]:
data.point_granularity.mode()
# output type is a series

0    1
dtype: int64

In [48]:
data.point_granularity.mode()[0]

1

In [49]:
data.point_granularity.median()

1.0

In [50]:
# Range of values in a column
(data.point_granularity.min(),data.point_granularity.max())

(1, 7)

In [55]:
data.point_granularity.count()

36634

In [53]:
data.county.dtype

dtype('O')

In [54]:
# How many non null values are there
data.county.count()
# Output type is an integer

36634

In [56]:
# For each uniwue value what is the count
data.county.value_counts()
# output type is a series
# In that series the labels are unique values
# the value for that label is how many times it occured in that column
# the final dtype refers to the value type of the series. count is integer

MIAMI DADE COUNTY    4315
BROWARD COUNTY       3193
PALM BEACH COUNTY    2791
DUVAL COUNTY         1894
ORANGE COUNTY        1811
                     ... 
GLADES COUNTY          22
UNION COUNTY           15
North Fort Myers        1
Orlando                 1
OSCEOLA COUNTY          1
Name: county, Length: 67, dtype: int64

In [59]:
# How many unique values are there
print(data.county.nunique())
print(data.county.value_counts().shape[0])

67
67


In [60]:
# What are the uinique values
print(data.county.unique())
print(data.county.value_counts().index)

['CLAY COUNTY' 'SUWANNEE COUNTY' 'NASSAU COUNTY' 'COLUMBIA COUNTY'
 'ST  JOHNS COUNTY' 'BAKER COUNTY' 'BRADFORD COUNTY' 'HAMILTON COUNTY'
 'UNION COUNTY' 'MADISON COUNTY' 'LAFAYETTE COUNTY' 'FLAGLER COUNTY'
 'DUVAL COUNTY' 'LAKE COUNTY' 'VOLUSIA COUNTY' 'PUTNAM COUNTY'
 'MARION COUNTY' 'SUMTER COUNTY' 'LEON COUNTY' 'FRANKLIN COUNTY'
 'LIBERTY COUNTY' 'GADSDEN COUNTY' 'WAKULLA COUNTY' 'JEFFERSON COUNTY'
 'TAYLOR COUNTY' 'BAY COUNTY' 'WALTON COUNTY' 'JACKSON COUNTY'
 'CALHOUN COUNTY' 'HOLMES COUNTY' 'WASHINGTON COUNTY' 'GULF COUNTY'
 'ESCAMBIA COUNTY' 'SANTA ROSA COUNTY' 'OKALOOSA COUNTY' 'ALACHUA COUNTY'
 'GILCHRIST COUNTY' 'LEVY COUNTY' 'DIXIE COUNTY' 'SEMINOLE COUNTY'
 'ORANGE COUNTY' 'BREVARD COUNTY' 'INDIAN RIVER COUNTY'
 'MIAMI DADE COUNTY' 'BROWARD COUNTY' 'MONROE COUNTY' 'PALM BEACH COUNTY'
 'MARTIN COUNTY' 'HENDRY COUNTY' 'PASCO COUNTY' 'GLADES COUNTY'
 'HILLSBOROUGH COUNTY' 'HERNANDO COUNTY' 'PINELLAS COUNTY' 'POLK COUNTY'
 'North Fort Myers' 'Orlando' 'HIGHLANDS COUNTY' 'HARDE

In [62]:
# To get the description for the numerical columns
data.describe()

policyID  eq_site_limit  hu_site_limit  fl_site_limit  \
count   36634.000000   3.663400e+04   3.663400e+04   3.663400e+04   
mean   548661.937189   7.314780e+05   2.074348e+06   6.646009e+05   
std    259912.926494   1.752300e+07   1.964150e+07   1.746114e+07   
min    100074.000000   0.000000e+00   0.000000e+00   0.000000e+00   
25%    323647.750000   0.000000e+00   4.248106e+04   0.000000e+00   
50%    548525.000000   0.000000e+00   1.926911e+05   0.000000e+00   
75%    774252.000000   0.000000e+00   9.163381e+05   0.000000e+00   
max    999971.000000   2.160000e+09   2.160000e+09   2.160000e+09   

       fr_site_limit      tiv_2011      tiv_2012  eq_site_deductible  \
count   3.663400e+04  3.663400e+04  3.663400e+04        3.663400e+04   
mean    9.911724e+05  2.172875e+06  2.571004e+06        7.787908e+02   
std     1.767178e+07  2.015015e+07  2.036777e+07        4.787979e+04   
min     0.000000e+00  9.000000e+01  7.337000e+01        0.000000e+00   
25%     0.000000e+00  4.562415e+04  5.401351e+04        0.000000e+00   
50%     0.000000e+00  2.021051e+05  2.416310e+05        0.000000e+00   
75%     5.741010e+04  9.465821e+05  1.128871e+06        0.000000e+00   
max     2.160000e+09  2.160000e+09  1.701000e+09        6.273765e+06   

       hu_site_deductible  fl_site_deductible  fr_site_deductible  \
count        3.663400e+04        36634.000000        36634.000000   
mean         7.037985e+03          192.453210           26.483594   
std          9.582593e+04         8037.713716         4702.740932   
min          0.000000e+00            0.000000            0.000000   
25%          0.000000e+00            0.000000            0.000000   
50%          0.000000e+00            0.000000            0.000000   
75%          0.000000e+00            0.000000            0.000000   
max          7.380000e+06       450000.000000       900000.000000   

       point_latitude  point_longitude  point_granularity  
count    36634.000000     36634.000000       36634.000000  
mean        28.087477       -81.903565           1.640907  
std          1.647734         1.809066           1.070561  
min         24.547514       -87.447290           1.000000  
25%         26.471996       -82.439331           1.000000  
50%         28.057077       -81.585700           1.000000  
75%         29.623005       -80.372910           3.000000  
max         30.989820       -80.033257           7.000000

In [63]:
# To get the list of numerical columns
data.describe().columns

Index(['policyID', 'eq_site_limit', 'hu_site_limit', 'fl_site_limit',
       'fr_site_limit', 'tiv_2011', 'tiv_2012', 'eq_site_deductible',
       'hu_site_deductible', 'fl_site_deductible', 'fr_site_deductible',
       'point_latitude', 'point_longitude', 'point_granularity'],
      dtype='object')

In [64]:
# To get the description for the categorical columns
data.describe(include=object)

statecode             county         line construction
count      36634              36634        36634        36634
unique         1                 67            2            5
top           FL  MIAMI DADE COUNTY  Residential         Wood
freq       36634               4315        30838        21581

In [66]:
data.describe(include=np.int64)

policyID  fr_site_deductible  point_granularity
count   36634.000000        36634.000000       36634.000000
mean   548661.937189           26.483594           1.640907
std    259912.926494         4702.740932           1.070561
min    100074.000000            0.000000           1.000000
25%    323647.750000            0.000000           1.000000
50%    548525.000000            0.000000           1.000000
75%    774252.000000            0.000000           3.000000
max    999971.000000       900000.000000           7.000000

In [67]:
data.describe(include=np.float64)

eq_site_limit  hu_site_limit  fl_site_limit  fr_site_limit  \
count   3.663400e+04   3.663400e+04   3.663400e+04   3.663400e+04   
mean    7.314780e+05   2.074348e+06   6.646009e+05   9.911724e+05   
std     1.752300e+07   1.964150e+07   1.746114e+07   1.767178e+07   
min     0.000000e+00   0.000000e+00   0.000000e+00   0.000000e+00   
25%     0.000000e+00   4.248106e+04   0.000000e+00   0.000000e+00   
50%     0.000000e+00   1.926911e+05   0.000000e+00   0.000000e+00   
75%     0.000000e+00   9.163381e+05   0.000000e+00   5.741010e+04   
max     2.160000e+09   2.160000e+09   2.160000e+09   2.160000e+09   

           tiv_2011      tiv_2012  eq_site_deductible  hu_site_deductible  \
count  3.663400e+04  3.663400e+04        3.663400e+04        3.663400e+04   
mean   2.172875e+06  2.571004e+06        7.787908e+02        7.037985e+03   
std    2.015015e+07  2.036777e+07        4.787979e+04        9.582593e+04   
min    9.000000e+01  7.337000e+01        0.000000e+00        0.000000e+00   
25%    4.562415e+04  5.401351e+04        0.000000e+00        0.000000e+00   
50%    2.021051e+05  2.416310e+05        0.000000e+00        0.000000e+00   
75%    9.465821e+05  1.128871e+06        0.000000e+00        0.000000e+00   
max    2.160000e+09  1.701000e+09        6.273765e+06        7.380000e+06   

       fl_site_deductible  point_latitude  point_longitude  
count        36634.000000    36634.000000     36634.000000  
mean           192.453210       28.087477       -81.903565  
std           8037.713716        1.647734         1.809066  
min              0.000000       24.547514       -87.447290  
25%              0.000000       26.471996       -82.439331  
50%              0.000000       28.057077       -81.585700  
75%              0.000000       29.623005       -80.372910  
max         450000.000000       30.989820       -80.033257

In [68]:
data.describe(include='all')

policyID statecode             county  eq_site_limit  \
count    36634.000000     36634              36634   3.663400e+04   
unique            NaN         1                 67            NaN   
top               NaN        FL  MIAMI DADE COUNTY            NaN   
freq              NaN     36634               4315            NaN   
mean    548661.937189       NaN                NaN   7.314780e+05   
std     259912.926494       NaN                NaN   1.752300e+07   
min     100074.000000       NaN                NaN   0.000000e+00   
25%     323647.750000       NaN                NaN   0.000000e+00   
50%     548525.000000       NaN                NaN   0.000000e+00   
75%     774252.000000       NaN                NaN   0.000000e+00   
max     999971.000000       NaN                NaN   2.160000e+09   

        hu_site_limit  fl_site_limit  fr_site_limit      tiv_2011  \
count    3.663400e+04   3.663400e+04   3.663400e+04  3.663400e+04   
unique            NaN            NaN            NaN           NaN   
top               NaN            NaN            NaN           NaN   
freq              NaN            NaN            NaN           NaN   
mean     2.074348e+06   6.646009e+05   9.911724e+05  2.172875e+06   
std      1.964150e+07   1.746114e+07   1.767178e+07  2.015015e+07   
min      0.000000e+00   0.000000e+00   0.000000e+00  9.000000e+01   
25%      4.248106e+04   0.000000e+00   0.000000e+00  4.562415e+04   
50%      1.926911e+05   0.000000e+00   0.000000e+00  2.021051e+05   
75%      9.163381e+05   0.000000e+00   5.741010e+04  9.465821e+05   
max      2.160000e+09   2.160000e+09   2.160000e+09  2.160000e+09   

            tiv_2012  eq_site_deductible  hu_site_deductible  \
count   3.663400e+04        3.663400e+04        3.663400e+04   
unique           NaN                 NaN                 NaN   
top              NaN                 NaN                 NaN   
freq             NaN                 NaN                 NaN   
mean    2.571004e+06        7.787908e+02        7.037985e+03   
std     2.036777e+07        4.787979e+04        9.582593e+04   
min     7.337000e+01        0.000000e+00        0.000000e+00   
25%     5.401351e+04        0.000000e+00        0.000000e+00   
50%     2.416310e+05        0.000000e+00        0.000000e+00   
75%     1.128871e+06        0.000000e+00        0.000000e+00   
max     1.701000e+09        6.273765e+06        7.380000e+06   

        fl_site_deductible  fr_site_deductible  point_latitude  \
count         36634.000000        36634.000000    36634.000000   
unique                 NaN                 NaN             NaN   
top                    NaN                 NaN             NaN   
freq                   NaN                 NaN             NaN   
mean            192.453210           26.483594       28.087477   
std            8037.713716         4702.740932        1.647734   
min               0.000000            0.000000       24.547514   
25%               0.000000            0.000000       26.471996   
50%               0.000000            0.000000       28.057077   
75%               0.000000            0.000000       29.623005   
max          450000.000000       900000.000000       30.989820   

        point_longitude         line construction  point_granularity  
count      36634.000000        36634        36634       36634.000000  
unique              NaN            2            5                NaN  
top                 NaN  Residential         Wood                NaN  
freq                NaN        30838        21581                NaN  
mean         -81.903565          NaN          NaN           1.640907  
std            1.809066          NaN          NaN           1.070561  
min          -87.447290          NaN          NaN           1.000000  
25%          -82.439331          NaN          NaN           1.000000  
50%          -81.585700          NaN          NaN           1.000000  
75%          -80.372910          NaN          NaN           3.000000  
max         

In [69]:
data.columns

Index(['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit',
       'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012',
       'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible',
       'fr_site_deductible', 'point_latitude', 'point_longitude', 'line',
       'construction', 'point_granularity'],
      dtype='object')

Slicing, subsetting, Boolean Indexing
- postion based (iloc)
    - rows --> 0 to 36633
    - columns --> 0 to 17
- label based indexing(loc)
    - rows --> 0 to 36633
    - columns --> ['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit',
       'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012',
       'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible',
       'fr_site_deductible', 'point_latitude', 'point_longitude', 'line',
       'construction', 'point_granularity']

In [70]:
data.iloc[1,:] # row with index 1 and all the columns

policyID                   448094
statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 1, dtype: object

In [72]:
data.loc[1,:]

policyID                   448094
statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 1, dtype: object

In [74]:
data.iloc[1,1] # row with index 1and column with index 1

'FL'

In [76]:
data.loc[1,"statecode"]

'FL'

In [78]:
data.iloc[1,1:3] # row with index 1 and columns with index 1,2
# iloc stop index is exclusive

statecode             FL
county       CLAY COUNTY
Name: 1, dtype: object

In [79]:
data.loc[1,"statecode":"county"] 
# loc - the stop label also inclusive

statecode             FL
county       CLAY COUNTY
Name: 1, dtype: object

In [80]:
data.iloc[1:4,1:3] 
# rows with indices 1,2,3
# columns with indices 1,2

statecode       county
1        FL  CLAY COUNTY
2        FL  CLAY COUNTY
3        FL  CLAY COUNTY

In [83]:
data.loc[1:3,"statecode":"county"] 

statecode       county
1        FL  CLAY COUNTY
2        FL  CLAY COUNTY
3        FL  CLAY COUNTY

In [84]:
data.policyID.nunique()

36634

In [85]:
data.set_index("policyID")

statecode           county  eq_site_limit  hu_site_limit  \
policyID                                                            
119736          FL      CLAY COUNTY       498960.0      498960.00   
448094          FL      CLAY COUNTY      1322376.3     1322376.30   
206893          FL      CLAY COUNTY       190724.4      190724.40   
333743          FL      CLAY COUNTY            0.0       79520.76   
172534          FL      CLAY COUNTY            0.0      254281.50   
...            ...              ...            ...            ...   
404309          FL  PINELLAS COUNTY            0.0       88722.00   
465875          FL  PINELLAS COUNTY      1297057.5     1297057.50   
791209          FL  PINELLAS COUNTY       173286.9      173286.90   
322627          FL  PINELLAS COUNTY      1499781.6     1499781.60   
398149          FL  PINELLAS COUNTY       373488.3      373488.30   

          fl_site_limit  fr_site_limit    tiv_2011    tiv_2012  \
policyID                                                         
119736         498960.0       498960.0   498960.00   792148.90   
448094        1322376.3      1322376.3  1322376.30  1438163.57   
206893         190724.4       190724.4   190724.40   192476.78   
333743              0.0            0.0    79520.76    86854.48   
172534              0.0       254281.5   254281.50   246144.49   
...                 ...            ...         ...         ...   
404309              0.0        88722.0    88722.00   155065.65   
465875        1297057.5      1297057.5  1297057.50  1577500.79   
791209              0.0            0.0   173286.90   205527.97   
322627              0.0            0.0  1499781.60  1768782.43   
398149              0.0            0.0   373488.30   596003.67   

          eq_site_deductible  hu_site_deductible  fl_site_deductible  \
policyID                                                               
119736                   0.0              9979.2                 0.0   
448094                   0.0                 0.0                 0.0   
206893                   0.0                 0.0                 0.0   
333743                   0.0                 0.0                 0.0   
172534                   0.0                 0.0                 0.0   
...                      ...                 ...                 ...   
404309                   0.0                 0.0                 0.0   
465875                   0.0                 0.0                 0.0   
791209                   0.0                 0.0                 0.0   
322627                   0.0                 0.0                 0.0   
398149                   0.0                 0.0                 0.0   

          fr_site_deductible  point_latitude  point_longitude         line  \
policyID                                                                     
119736                     0       30.102261       -81.711777  Residential   
448094                     0       30.063936       -81.707664  Residential   
206893                     0       30.089579       -81.700455  Residential   
333743                     0       30.063236       -81.707703  Residential   
172534                     0       30.060614       -81.702675  Residential   
...                      ...             ...              ...          ...   
404309                     0       28.122885       -82.770218  Residential   
465875                     0       28.080900       -82.758800  Residential   
791209                     0       28.110550       -82.766360  Residential   
322627                     0       28.089415       -82.697411  Residential   
398149                     0       28.064440       -82.774590  Residential   

         construction  point_granularity  
policyID                                  
119736        Masonry                  1  
448094        Masonry                  3  
206893           Wood                  1  
333743           Wood                  3  
172534           Wood                  1  

In [86]:
data

policyID statecode           county  eq_site_limit  hu_site_limit  \
0        119736        FL      CLAY COUNTY       498960.0      498960.00   
1        448094        FL      CLAY COUNTY      1322376.3     1322376.30   
2        206893        FL      CLAY COUNTY       190724.4      190724.40   
3        333743        FL      CLAY COUNTY            0.0       79520.76   
4        172534        FL      CLAY COUNTY            0.0      254281.50   
...         ...       ...              ...            ...            ...   
36629    404309        FL  PINELLAS COUNTY            0.0       88722.00   
36630    465875        FL  PINELLAS COUNTY      1297057.5     1297057.50   
36631    791209        FL  PINELLAS COUNTY       173286.9      173286.90   
36632    322627        FL  PINELLAS COUNTY      1499781.6     1499781.60   
36633    398149        FL  PINELLAS COUNTY       373488.3      373488.30   

       fl_site_limit  fr_site_limit    tiv_2011    tiv_2012  \
0           498960.0       498960.0   498960.00   792148.90   
1          1322376.3      1322376.3  1322376.30  1438163.57   
2           190724.4       190724.4   190724.40   192476.78   
3                0.0            0.0    79520.76    86854.48   
4                0.0       254281.5   254281.50   246144.49   
...              ...            ...         ...         ...   
36629            0.0        88722.0    88722.00   155065.65   
36630      1297057.5      1297057.5  1297057.50  1577500.79   
36631            0.0            0.0   173286.90   205527.97   
36632            0.0            0.0  1499781.60  1768782.43   
36633            0.0            0.0   373488.30   596003.67   

       eq_site_deductible  hu_site_deductible  fl_site_deductible  \
0                     0.0              9979.2                 0.0   
1                     0.0                 0.0                 0.0   
2                     0.0                 0.0                 0.0   
3                     0.0                 0.0                 0.0   
4                     0.0                 0.0                 0.0   
...                   ...                 ...                 ...   
36629                 0.0                 0.0                 0.0   
36630                 0.0                 0.0                 0.0   
36631                 0.0                 0.0                 0.0   
36632                 0.0                 0.0                 0.0   
36633                 0.0                 0.0                 0.0   

       fr_site_deductible  point_latitude  point_longitude         line  \
0                       0       30.102261       -81.711777  Residential   
1                       0       30.063936       -81.707664  Residential   
2                       0       30.089579       -81.700455  Residential   
3                       0       30.063236       -81.707703  Residential   
4                       0       30.060614       -81.702675  Residential   
...                   ...             ...              ...          ...   
36629                   0       28.122885       -82.770218  Residential   
36630                   0       28.080900       -82.758800  Residential   
36631                   0       28.110550       -82.766360  Residential   
36632                   0       28.089415       -82.697411  Residential   
36633                   0       28.064440       -82.774590  Residential   

      construction  point_granularity  
0          Masonry                  1  
1          Masonry                  3  
2             Wood                  1  
3             Wood                  3  
4             Wood                  1  
...            ...                ...  
36629         Wood                  3  
36630      Masonry                  1  
36631         Wood                  4  
36632      Masonry                  3  
36633      Masonry                  1  

[36634 rows x 18 columns]

In [87]:
df = data.set_index("policyID")
df.head()

statecode       county  eq_site_limit  hu_site_limit  fl_site_limit  \
policyID                                                                       
119736          FL  CLAY COUNTY       498960.0      498960.00       498960.0   
448094          FL  CLAY COUNTY      1322376.3     1322376.30      1322376.3   
206893          FL  CLAY COUNTY       190724.4      190724.40       190724.4   
333743          FL  CLAY COUNTY            0.0       79520.76            0.0   
172534          FL  CLAY COUNTY            0.0      254281.50            0.0   

          fr_site_limit    tiv_2011    tiv_2012  eq_site_deductible  \
policyID                                                              
119736         498960.0   498960.00   792148.90                 0.0   
448094        1322376.3  1322376.30  1438163.57                 0.0   
206893         190724.4   190724.40   192476.78                 0.0   
333743              0.0    79520.76    86854.48                 0.0   
172534         254281.5   254281.50   246144.49                 0.0   

          hu_site_deductible  fl_site_deductible  fr_site_deductible  \
policyID                                                               
119736                9979.2                 0.0                   0   
448094                   0.0                 0.0                   0   
206893                   0.0                 0.0                   0   
333743                   0.0                 0.0                   0   
172534                   0.0                 0.0                   0   

          point_latitude  point_longitude         line construction  \
policyID                                                              
119736         30.102261       -81.711777  Residential      Masonry   
448094         30.063936       -81.707664  Residential      Masonry   
206893         30.089579       -81.700455  Residential         Wood   
333743         30.063236       -81.707703  Residential         Wood   
172534         30.060614       -81.702675  Residential         Wood   

          point_granularity  
policyID                     
119736                    1  
448094                    3  
206893                    1  
333743                    3  
172534                    1

In [88]:
df.index

Int64Index([119736, 448094, 206893, 333743, 172534, 785275, 995932, 223488,
            433512, 142071,
            ...
            827844, 876385, 167630, 241496, 780814, 404309, 465875, 791209,
            322627, 398149],
           dtype='int64', name='policyID', length=36634)

Slicing, subsetting, Boolean Indexing with df
- postion based (iloc)
    - rows --> 0 to 36633
    - columns --> 0 to 17
- label based indexing(loc)
    - rows --> [119736, 448094, 206893, 333743, 172534, 785275, 995932, 223488,
            433512, 142071,...]
    - columns --> ['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit',
       'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012',
       'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible',
       'fr_site_deductible', 'point_latitude', 'point_longitude', 'line',
       'construction', 'point_granularity']

In [ ]:
# row with index 1 and all the columns

In [ ]:
# row with index 1and column with index 1

In [ ]:
# row with index 1 and columns with index 1,2

In [ ]:
# rows with indices 1,2,3
# columns with indices 1,2